aten/src/ATen/core/op_registration/kernel_functor.h

```diff
@@ -70,9 +70,31 @@ namespace detail {
     }
   };
   template<class T>
+  struct ivalue_to_arg_type<ListPtr<T>, guts::enable_if_t<guts::typelist::contains<supported_primitive_arg_types, T>::value && !std::is_same<std::string, T>::value>> {
+    static ListPtr<T> call(IValue&& v) {
+      static_assert(guts::typelist::contains<supported_primitive_arg_types, T>::value, "You tried to register a kernel with an unsupported argument type: c10::List<T> and T is not one of the supported primitive types.");
+
+      return std::move(*std::move(v).to<intrusive_ptr<ivalue::List<T>>>()).elements();
+    }
+  };
+  template<class T>
+  struct ivalue_to_arg_type<ListPtr<T>, guts::enable_if_t<!guts::typelist::contains<supported_primitive_arg_types, T>::value || std::is_same<std::string, T>::value>> {
+    static ListPtr<T> call(IValue&& v) {
+      static_assert(guts::typelist::contains<supported_primitive_arg_types, T>::value, "You tried to register a kernel with an unsupported argument type: c10::List<T> and T is not one of the supported primitive types.");
```

Comment:
we might want to allow `List<IValue>` actually, why explicitly forbid it? (i.e. what if it's a list of tuples)

Reply:
Using explicit types is a much nicer API. If there's a use case for Tuples, we should add the `TuplePtr` type here. I want kernel writers to use IValue as little as possible.